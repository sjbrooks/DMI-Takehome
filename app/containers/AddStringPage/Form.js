import styled from 'styled-components';

const Form = styled.form`
  margin: 2rem 0;
`;

export default Form;
