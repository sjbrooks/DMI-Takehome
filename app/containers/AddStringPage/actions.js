/*
 * Add String Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  CHANGE_STRING,
  CREATE_STRING,
  CREATE_STRING_SUCCESS,
  CREATE_STRING_ERROR,
} from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {string} string The new text of the input field
 *
 * @return {object} An action object with a type of CHANGE_STRING
 */
export function changeString(string) {
  return {
    type: CHANGE_STRING,
    string,
  };
}

/**
 * Create a new string, this action starts the request saga
 *
 * @return {object} An action object with a type of CREATE_STRING
 */
export function createString(string) {
  return {
    type: CREATE_STRING,
    string,
  };
}

/**
 * Dispatched when the string has been created by the request saga
 *
 * @param  {array} repos The repository data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_STRINGS_SUCCESS passing the repos
 */
export function stringCreated(string) {
  return {
    type: CREATE_STRING_SUCCESS,
    string,
  };
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_STRINGS_ERROR passing the error
 */
export function stringCreationError(error) {
  return {
    type: CREATE_STRING_ERROR,
    error,
  };
}
