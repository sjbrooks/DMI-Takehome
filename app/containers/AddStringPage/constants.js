/*
 * AddStringPageConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const CHANGE_STRING = 'DMI-Takehome/AddStringPage/CHANGE_STRING';
export const CREATE_STRING = 'DMI-Takehome/AddStringPage/CREATE_STRING';
export const CREATE_STRING_SUCCESS =
  'DMI-Takehome/AddStringPage/CREATE_STRING_SUCCESS';
export const CREATE_STRING_ERROR =
  'DMI-Takehome/AddStringPage/CREATE_STRING_ERROR';
