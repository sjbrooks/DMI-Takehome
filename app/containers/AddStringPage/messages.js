/*
 * FeaturePage Messages
 *
 * This contains all the text for the FeaturePage component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'boilerplate.containers.FeaturePage';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'Add a string',
  },
});
