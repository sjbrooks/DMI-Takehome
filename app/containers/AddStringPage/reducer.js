/*
 * addStringReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import {
  CHANGE_STRING,
  CREATE_STRING,
  CREATE_STRING_SUCCESS,
  CREATE_STRING_ERROR,
} from './constants';

// The initial state of the AddStringPage
export const initialState = {
  loading: false,
  error: false,
  data: {
    string: '',
  },
  created: false,
};

/* eslint-disable default-case, no-param-reassign */
const addStringReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_STRING:
        draft.data.string = action.string;
        draft.error = false;
        draft.created = false;
        break;

      case CREATE_STRING:
        draft.loading = false;
        draft.error = false;
        draft.created = false;
        break;

      case CREATE_STRING_SUCCESS:
        draft.loading = false;
        draft.created = true;
        break;

      case CREATE_STRING_ERROR:
        draft.error = action.error;
        draft.loading = false;
        draft.created = false;
        break;
    }
  });

export default addStringReducer;
