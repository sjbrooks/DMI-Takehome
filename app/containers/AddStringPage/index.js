/*
 * AddStringPage
 *
 * List all the features
 */
import React from 'react';
import { Helmet } from 'react-helmet';
import { FormattedMessage } from 'react-intl';
import H1 from 'components/H1';
// import Form from './Form';
// import Input from './Input';
// import { changeString } from './actions';
// import { makeSelectUsername } from './selectors';

import messages from './messages';

export default function AddStringPage() {
  return (
    <div>
      <Helmet>
        <title>Add String Page</title>
        <meta
          name="description"
          content="Form to add a string to string store application"
        />
      </Helmet>
      <H1>
        <FormattedMessage {...messages.header} />
      </H1>
      {/* <Form onSubmit={onSubmitForm}>
        <label htmlFor="username">
          <FormattedMessage {...messages.trymeMessage} />
          <AtPrefix>
            <FormattedMessage {...messages.trymeAtPrefix} />
          </AtPrefix>
          <Input
            id="string"
            type="text"
            placeholder="mxstbr"
            value={string}
            onChange={onChangeString}
          />
        </label>
      </Form> */}
    </div>
  );
}
