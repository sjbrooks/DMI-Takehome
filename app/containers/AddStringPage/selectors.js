/**
 * AddStringPage selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectAddStringPage = state => state.addString || initialState;

const makeSelectString = () =>
  createSelector(
    selectAddStringPage,
    addStringState => addStringState.data.string,
  );

export { selectAddStringPage, makeSelectString };
