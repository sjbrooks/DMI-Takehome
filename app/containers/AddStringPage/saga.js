/**
 * Posts the new string to fakeDB
 */

import { call, put, select, takeLatest } from 'redux-saga/effects';
import { CREATE_STRING } from 'containers/AddStringPage/constants';
import {
  stringCreated,
  stringCreationError,
} from 'containers/AddStringPage/actions';

import request from 'utils/request';
import { makeSelectString } from 'containers/AddStringPage/selectors';

/**
 * Post new string request/response handler
 */
export function* createString() {
  // Select current string in input field from store
  let string = yield select(makeSelectString());
  const requestURL = `http://localhost:3000/strings`;
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ string }),
  };

  try {
    // Prevent empty form from being submitted
    if (string !== '') {
      // Call our request helper (see 'utils/request')
      string = yield call(request, requestURL, options);
      yield put(stringCreated(string));
    }
  } catch (err) {
    yield put(stringCreationError(err));
  }
}

/**
 * AddStringPage saga manages watcher lifecycle
 */
export default function* newString() {
  // Watches for CREATE_STRING actions and calls createString when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(CREATE_STRING, createString);
}
