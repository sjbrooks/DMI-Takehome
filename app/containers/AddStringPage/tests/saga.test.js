/**
 * Tests for AddStringPage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { CREATE_STRING } from 'containers/AddStringPage/constants';
import {
  stringCreated,
  stringCreationError,
} from 'containers/AddStringPage/actions';

import newString, { createString } from '../saga';

const string = 'test string';

/* eslint-disable redux-saga/yield-effects */
describe('createString Saga', () => {
  let createStringGenerator;

  // We have to test twice, once for a successful string creation and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    createStringGenerator = createString();

    const selectDescriptor = createStringGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = createStringGenerator.next(string).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the stringCreated action if it creates the string successfully', () => {
    const response = string;

    const putDescriptor = createStringGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(stringCreated(string)));
  });

  it('should call the stringCreationError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = createStringGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(stringCreationError(response)));
  });
});

describe('newStringSaga Saga', () => {
  const newStringSaga = newString();

  it('should start task to watch for CREATE_STRING action', () => {
    const takeLatestDescriptor = newStringSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(CREATE_STRING, createString),
    );
  });
});
