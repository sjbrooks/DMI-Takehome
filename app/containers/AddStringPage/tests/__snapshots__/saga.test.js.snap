// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createString Saga should call the stringCreationError action if string is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createString Saga should call the stringCreationError action if string is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/strings",
      Object {
        "body": "{\\"string\\":\\"test string\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`createString Saga should call the stringCreationError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createString Saga should call the stringCreationError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/strings",
      Object {
        "body": "{\\"string\\":\\"test string\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`createString Saga should dispatch the stringCreated action if it creates the string successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createString Saga should dispatch the stringCreated action if it creates the string successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/strings",
      Object {
        "body": "{\\"string\\":\\"test string\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`createString Saga should not dispatch the stringCreated action if string is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createString Saga should not dispatch the stringCreated action if string is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/strings",
      Object {
        "body": "{\\"string\\":\\"test string\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
